File: 89.js
function subsets(nums) {
    const res = [];
    const subset = [];

    function backtrack(i) {
        if (i === nums.length) {
            res.push([...subset]);
            return;
        }
        subset.push(nums[i]);
        backtrack(i + 1);
        subset.pop();
        backtrack(i + 1);
    }

    backtrack(0);
    return res;
}

console.log(subsets([1,2]));

File: 89.py
def subsets(nums):
    res = []
    subset = []

    def backtrack(i):
        if i == len(nums):
            res.append(subset.copy())
            return
        # include nums[i]
        subset.append(nums[i])
        backtrack(i + 1)
        subset.pop()
        # exclude nums[i]
        backtrack(i + 1)

    backtrack(0)
    return res

print(subsets([1,2]))
